apps/emqx_message_transformation/src/emqx_message_transformation.erl

```diff
@@ -0,0 +1,820 @@
+        error ->
+            FailureAction
+    end.
+
+take_from_context(Context, Message) ->
+    maps:fold(
+        fun
+            (payload, _, Acc) ->
+                Acc#message{payload = maps:get(payload, Context)};
+            (qos, _, Acc) ->
+                Acc#message{qos = maps:get(qos, Context)};
+            (topic, _, Acc) ->
+                Acc#message{topic = maps:get(topic, Context)};
+            (retain, _, Acc) ->
+                emqx_message:set_flag(retain, maps:get(retain, Context), Acc);
+            (user_property, _, Acc) ->
+                Props0 = emqx_message:get_header(properties, Acc, #{}),
+                Props = maps:merge(Props0, #{'User-Property' => maps:get(user_property, Context)}),
+                emqx_message:set_header(properties, Props, Acc)
+        end,
+        Message,
+        maps:get(dirty, Context)
+    ).
+
+decode(Payload, #{type := none}, _Transformation) ->
+    {ok, Payload};
+decode(Payload, #{type := json}, Transformation) ->
+    case emqx_utils_json:safe_decode(Payload, [return_maps]) of
+        {ok, JSON} ->
+            {ok, JSON};
+        {error, Reason} ->
+            trace_failure(Transformation, "payload_decode_failed", #{
+                decoder => json,
+                reason => Reason
+            }),
+            error
+    end;
+decode(Payload, #{type := avro, schema := SerdeName}, Transformation) ->
+    try
+        {ok, emqx_schema_registry_serde:decode(SerdeName, Payload)}
+    catch
+        error:{serde_not_found, _} ->
+            trace_failure(Transformation, "payload_decode_schema_not_found", #{
+                decoder => avro,
+                schema_name => SerdeName
+            }),
+            error;
+        Class:Error:Stacktrace ->
+            trace_failure(Transformation, "payload_decode_schema_failure", #{
+                decoder => avro,
+                schema_name => SerdeName,
+                kind => Class,
+                reason => Error,
+                stacktrace => Stacktrace
+            }),
+            error
+    end;
+decode(
+    Payload, #{type := protobuf, schema := SerdeName, message_type := MessageType}, Transformation
+) ->
+    try
+        {ok, emqx_schema_registry_serde:decode(SerdeName, Payload, [MessageType])}
+    catch
+        error:{serde_not_found, _} ->
+            trace_failure(Transformation, "payload_decode_schema_not_found", #{
+                decoder => protobuf,
+                schema_name => SerdeName,
+                message_type => MessageType
+            }),
+            error;
+        Class:Error:Stacktrace ->
+            trace_failure(Transformation, "payload_decode_schema_failure", #{
+                decoder => protobuf,
+                schema_name => SerdeName,
+                message_type => MessageType,
+                kind => Class,
+                reason => Error,
+                stacktrace => Stacktrace
+            }),
+            error
+    end.
```

Comment:
Nit: could we squash `emqx_schema_registry_serde` cases to a single one to avoid implementing parallel dispatching for each schema registry's module?

Reply:
Currently, we have only 2: avro and protobuf.  We can't squash those because only protobuf has `message_type`, which is a required extra argument for the serde.